File: services/ml-service/app/backtest.py
"""Walk-forward backtester for the ML model — the honesty gate (ADR-0004/0005).

Same causal methodology as the TS backtester: at each step we train ONLY on data
strictly before the prediction day, predict the next day, and score against the
realized outcome. We compare the ML model to naive random-walk and always-up
baselines on the identical window. This retrains repeatedly (expanding window with a
refit cadence) — the ONLY honest way to test a learned model. Scoring on training
data would be a lie.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
from lightgbm import LGBMClassifier

from .config import BAND_LOWER_Q, BAND_UPPER_Q
from .features import FEATURE_COLUMNS, build_features, make_labels
from .model import _COMMON


def walk_forward(
    closes: pd.Series,
    horizon: int = 1,
    min_train: int = 250,
    test_days: int = 250,
    refit_every: int = 10,
) -> dict:
    """Expanding-window walk-forward. Returns honest metrics vs baselines.

    refit_every: retrain cadence. Refitting every single day is ideal but slow;
    every 10 days is a well-accepted approximation that does NOT leak future data
    (we still only ever train on the past).
    """
    feats_all = build_features(closes)
    fwd_all, up_all = make_labels(closes, horizon)

    df = feats_all.copy()
    df["_fwd"] = fwd_all
    df["_up"] = up_all
    df = df.dropna()

    n = len(df)
    if n < min_train + 30:
        raise ValueError(f"insufficient_history: have {n}, need {min_train + 30}")

    start = max(min_train, n - test_days)

    model_correct = naive_correct = up_correct = total = 0
    band_hits = 0
    abs_err = []
    prev_close_col = closes.reindex(df.index)

    clf = None
    for i in range(start, n):
        # Train window: strictly rows before i.
        if clf is None or (i - start) % refit_every == 0:
            X_tr = df[FEATURE_COLUMNS].iloc[:i]
            y_tr = df["_up"].iloc[:i]
            clf = LGBMClassifier(**_COMMON)
            clf.fit(X_tr, y_tr)

        x_i = df[FEATURE_COLUMNS].iloc[[i]]
        actual_up = df["_up"].iloc[i] >= 0.5
        pred_up = clf.predict_proba(x_i)[0, 1] >= 0.5
        if pred_up == actual_up:
            model_correct += 1

        # Baselines on the same row.
        # naive: same direction as the last observed move (ret_1 sign).
        naive_up = df["ret_1"].iloc[i] >= 0
        if naive_up == actual_up:
            naive_correct += 1
        if actual_up:
            up_correct += 1

        total += 1

    hit = model_correct / total if total else 0.0
    naive_hit = naive_correct / total if total else 0.0
    up_hit = up_correct / total if total else 0.0

    return {
        "horizon": horizon,
        "samples": total,
        "refit_every": refit_every,
        "model_hit_rate": round(hit, 4),
        "naive_hit_rate": round(naive_hit, 4),
        "always_up_hit_rate": round(up_hit, 4),
        "edge_vs_naive_pts": round((hit - naive_hit) * 100, 2),
        "edge_vs_up_pts": round((hit - up_hit) * 100, 2),
        "date_from": str(df.index[start]),
        "date_to": str(df.index[-1]),
        "verdict": _verdict(hit, naive_hit, up_hit, total),
    }


def _verdict(hit: float, naive: float, up: float, n: int) -> str:
    parts = []
    if n < 100:
        parts.append(f"Only {n} samples — indicative, not conclusive.")
    e_naive = (hit - naive) * 100
    e_up = (hit - up) * 100
    parts.append(f"ML {hit*100:.1f}% vs naive {naive*100:.1f}% ({e_naive:+.1f} pts), vs always-up {up*100:.1f}% ({e_up:+.1f} pts).")
    best_baseline = max(naive, up)
    if hit > best_baseline + 0.02:
        parts.append("ML beats BOTH baselines — a genuine edge; worth shipping.")
    elif hit > best_baseline:
        parts.append("ML edges out the baselines slightly — marginal; keep watching.")
    else:
        parts.append("ML does NOT beat the best baseline — honest result: do not ship it as 'better'. Keep the simple engine.")
    return " ".join(parts)

File: apps/web/src/lib/engine/backtest.ts
import type { AssetId, MacroSnapshot, PricePoint } from "@/lib/domain/types";
import type { IAnalysisEngine } from "@/lib/engine/types";

/**
 * Walk-forward backtester — the honesty engine of Phoenix (ADR-0004).
 *
 * At each test day t, we hand the engine ONLY the price history strictly before the
 * outcome we're scoring. It predicts the next day's direction; we compare to what
 * actually happened. There is NO way for future data to leak in: we slice the array
 * up to t and pass that slice. This is what stops us fooling ourselves.
 *
 * We score the model against two baselines:
 *   - naive random walk: predict "same direction as the last move"
 *   - always-up: gold drifts up long-term; a hard baseline to beat on hit rate
 */

export interface BacktestConfig {
  asset: AssetId;
  /** How many of the most recent days to evaluate. */
  testDays: number;
  /** Minimum history the engine needs before its first prediction. */
  minTrain: number;
  /** Forecast horizon in days (1 = next day). */
  horizonDays: number;
}

export interface BacktestResult {
  asset: AssetId;
  horizonDays: number;
  samples: number;
  dateRange: { from: string; to: string };
  model: DirectionalScore & ErrorScore & { bandCoverage: number; bandTarget: number };
  baselineNaive: DirectionalScore;
  baselineAlwaysUp: DirectionalScore;
  /** Plain-language verdict — honest, never inflated. */
  verdict: string;
}

interface DirectionalScore {
  hitRate: number; // 0..1
  correct: number;
  total: number;
}
interface ErrorScore {
  mae: number; // mean absolute error of central estimate ($)
  rmse: number;
}

export async function backtest(
  engine: IAnalysisEngine,
  fullSeries: PricePoint[],
  macro: MacroSnapshot,
  cfg: BacktestConfig,
): Promise<BacktestResult> {
  const { asset, horizonDays } = cfg;
  const n = fullSeries.length;

  // Determine the evaluation window [start, end) of OUTCOME indices.
  // We predict series[i] using only series[0..i-1] (strictly causal).
  const firstOutcome = Math.max(cfg.minTrain, 1);
  const lastOutcome = n - 1; // last index we have a real outcome for
  const startOutcome = Math.max(firstOutcome, lastOutcome - cfg.testDays + 1);

  let modelCorrect = 0;
  let naiveCorrect = 0;
  let upCorrect = 0;
  let total = 0;
  let absErrSum = 0;
  let sqErrSum = 0;
  let bandHits = 0;
  let bandTarget = 0.8;

  for (let i = startOutcome; i <= lastOutcome; i++) {
    // History available to the model: everything strictly before the outcome day i,
    // ending at the "as of" day (i - horizonDays) ... for horizon 1 that's i-1.
    const asOfIdx = i - horizonDays;
    if (asOfIdx < cfg.minTrain - 1) continue;

    const history = fullSeries.slice(0, asOfIdx + 1); // inclusive of as-of day
    const asOfPrice = history[history.length - 1].close;
    const actual = fullSeries[i].close;
    const actualUp = actual >= asOfPrice;

    // --- Model prediction (only past data) ---
    const result = await engine.analyze({ asset, series: history, macro, horizonDays });
    const predUp = result.forecast.probUp >= 0.5;
    if (predUp === actualUp) modelCorrect++;

    // central-estimate error + band coverage (calibration)
    absErrSum += Math.abs(result.forecast.central - actual);
    sqErrSum += (result.forecast.central - actual) ** 2;
    bandTarget = result.forecast.intervalCoverage;
    if (actual >= result.forecast.lower && actual <= result.forecast.upper) bandHits++;

    // --- Baselines ---
    // naive: predict same direction as the last observed move
    if (history.length >= 2) {
      const prev = history[history.length - 2].close;
      const naiveUp = asOfPrice >= prev;
      if (naiveUp === actualUp) naiveCorrect++;
    }
    // always-up
    if (actualUp) upCorrect++;

    total++;
  }

  const model = {
    hitRate: safeDiv(modelCorrect, total),
    correct: modelCorrect,
    total,
    mae: round2(safeDiv(absErrSum, total)),
    rmse: round2(Math.sqrt(safeDiv(sqErrSum, total))),
    bandCoverage: round3(safeDiv(bandHits, total)),
    bandTarget,
  };

  const result: BacktestResult = {
    asset,
    horizonDays,
    samples: total,
    dateRange: {
      from: fullSeries[startOutcome]?.date ?? "n/a",
      to: fullSeries[lastOutcome]?.date ?? "n/a",
    },
    model,
    baselineNaive: { hitRate: safeDiv(naiveCorrect, total), correct: naiveCorrect, total },
    baselineAlwaysUp: { hitRate: safeDiv(upCorrect, total), correct: upCorrect, total },
    verdict: buildVerdict(model.hitRate, safeDiv(naiveCorrect, total), total, model.bandCoverage, bandTarget),
  };
  return result;
}

/** Honest, non-inflated plain-language read of the numbers (ADR-0004 §5). */
function buildVerdict(
  modelHit: number,
  naiveHit: number,
  samples: number,
  bandCov: number,
  bandTarget: number,
): string {
  const parts: string[] = [];
  const edge = modelHit - naiveHit;
  const pctM = (modelHit * 100).toFixed(1);
  const pctN = (naiveHit * 100).toFixed(1);

  if (samples < 60) {
    parts.push(`Only ${samples} samples — too few to claim skill; treat as indicative.`);
  }
  if (edge > 0.02) {
    parts.push(`Model direction ${pctM}% beats naive ${pctN}% by ${(edge * 100).toFixed(1)} pts — a genuine but modest edge.`);
  } else if (edge < -0.02) {
    parts.push(`Model ${pctM}% is BELOW naive ${pctN}% — the model is not adding directional value here. Honest result: don't trust the direction yet.`);
  } else {
    parts.push(`Model ${pctM}% ≈ naive ${pctN}% — essentially a coin flip, as expected for next-day gold. The value is in the probability + band, not the call.`);
  }

  const covGap = Math.abs(bandCov - bandTarget);
  if (covGap <= 0.1) {
    parts.push(`Band is well-calibrated (${(bandCov * 100).toFixed(0)}% actual vs ${(bandTarget * 100).toFixed(0)}% target).`);
  } else if (bandCov < bandTarget) {
    parts.push(`Band is too tight (${(bandCov * 100).toFixed(0)}% actual vs ${(bandTarget * 100).toFixed(0)}% target) — it overstates confidence.`);
  } else {
    parts.push(`Band is conservative (${(bandCov * 100).toFixed(0)}% actual vs ${(bandTarget * 100).toFixed(0)}% target) — wider than needed, but honest.`);
  }
  return parts.join(" ");
}

function safeDiv(a: number, b: number): number {
  return b === 0 ? 0 : a / b;
}
function round2(x: number): number {
  return Math.round(x * 100) / 100;
}
function round3(x: number): number {
  return Math.round(x * 1000) / 1000;
}
